web/src/engine/websites/ZenManga.ts

```diff
@@ -0,0 +1,93 @@
+﻿import { Tags } from '../Tags';
+import icon from './ZenManga.webp';
+import { Chapter, DecoratableMangaScraper, Manga, type MangaPlugin, Page } from '../providers/MangaPlugin';
+import * as Common from './decorators/Common';
+import { FetchCSS, FetchJSON } from '../platform/FetchProvider';
+import * as devalue from 'devalue';
+
+type APIManga = {
+    id: string;
+    slug: string;
+    name: {
+        ru: string;
+    }
+};
+
+type HydratedChapter = {
+    id: string;
+    name: string;
+    number: number;
+    branchId: string;
+    volume: number;
+};
+
+type HydratedBranch = {
+    id: string;
+    publishers: {
+        name: string;
+    }[]
+}
+
+type APIPages = {
+    pages: {
+        image: string;
+    }[]
+};
+
+@Common.ImageAjax()
+export default class extends DecoratableMangaScraper {
+    private readonly apiUrl = 'https://api.zenmanga.io/v2/';
+
+    public constructor() {
+        super('zenmanga', 'ZenManga', 'https://zenmanga.io', Tags.Media.Manhwa, Tags.Media.Manhua, Tags.Language.Russian, Tags.Source.Aggregator);
+    }
+
+    public override get Icon() {
+        return icon;
+    }
+
+    public override ValidateMangaURL(url: string): boolean {
+        return new RegExpSafe(`^${this.URI.origin}/content/[^/]+$`).test(url);
+    }
+
+    public override async FetchManga(provider: MangaPlugin, url: string): Promise<Manga> {
+        const { slug, name: { ru } } = await FetchJSON<APIManga>(new Request(new URL(`./books/${url.split('/').at(-1)}`, this.apiUrl)));
+        return new Manga(this, provider, slug, ru);
+    }
+
+    public override async FetchMangas(provider: MangaPlugin): Promise<Manga[]> {
+        const mangaList: Manga[] = [];
+        for (let page = 0, run = true; run; page++) {
+            const mangas = await this.GetMangasFromPage(page, provider);
+            mangas.length > 0 ? mangaList.push(...mangas) : run = false;
+        }
+        return mangaList;
+    }
+
+    private async GetMangasFromPage(page: number, provider: MangaPlugin): Promise<Manga[]> {
+        const data = await FetchJSON<APIManga[]>(new Request(new URL(`./books?page=${page}&size=200&sort=viewsCount%2Cdesc`, this.apiUrl)));
+        return data.map(({ slug, name: { ru } }) => new Manga(this, provider, slug, ru));
+    }
+
+    public override async FetchChapters(manga: Manga): Promise<Chapter[]> {
+        const revivers = {
+            URL: t => new URL(t),
+            Date: t => new Date(t),
+        };
+
+        const [script] = await FetchCSS<HTMLScriptElement>(new Request(new URL(`./content/${manga.Identifier}`, this.URI)), 'script#it-astro-state');
```

Comment:
Consider to let the website doing the state extraction and just get the data which is necessary via `FetchJavaScript` (or with preload script if necessary)
```javascript
new Promise(resolve => {
    document.addEventListener('@it-astro:server-state-loaded', event => {
        const chapters = event.serverState
            .get('@inox-tools/request-nanostores')
            .get('current-book-chapters');
        resolve(chapters);
    });
});
```

Reply:
The event is probably added to late, use preload fetch

Reply:
yeah figured it out. Gotta listen to that event, then dispatch a custom event with relevant data so its catched by main script